content_scripts/mode.coffee

```diff
@@ -0,0 +1,210 @@
+# Modes.
+#
+# A mode implements a number of keyboard event handlers which are pushed onto the handler stack when the mode
+# starts, and poped when the mode exits.  The Mode base class takes as single argument options which can
+# define:
+#
+# name:
+#   A name for this mode.
+#
+# badge:
+#   A badge (to appear on the browser popup) for this mode.
+#   Optional.  Define a badge is the badge is constant.  Otherwise, do not define a badge and override the
+#   chooseBadge method instead.  Or, if the mode *never* shows a badge, then do neither.
+#
+# keydown:
+# keypress:
+# keyup:
+#   Key handlers.  Optional: provide these as required.  The default is to continue bubbling all key events.
+#
+# Additional handlers associated with the mode can be added by using the push method.  For example, if a mode
+# responds to "focus" events, then push an additional handler:
+#   @push
+#     "focus": (event) => ....
+# Any such additional handlers are removed when the mode exits.
+#
+# New mode types are created by inheriting from Mode or one of its sub-classes.  Some generic cub-classes are
+# provided below:
+#
+#   SingletonMode: ensures that at most one instance of the mode is active at any one time.
+#   ConstrainedMode: exits the mode if the an indicated element loses the focus.
+#   ExitOnEscapeMode: exits the mode on escape.
+#   StateMode: tracks the current Vimium state in @enabled and @passKeys.
+#
+# To install and existing mode, use:
+#   myMode = new MyMode()
+#
+# To remove a mode, use:
+#   myMode.exit() # externally triggered.
+#   @exit()       # internally triggered (more common).
+#
+
+# For debug only; to be stripped out.
+count = 0
+
+class Mode
+  # Static.
+  @modes: []
+
+  # Constants; readable shortcuts for event-handler return values.
+  continueBubbling: true
+  suppressEvent: false
+  stopBubblingAndTrue: handlerStack.stopBubblingAndTrue
+  stopBubblingAndFalse: handlerStack.stopBubblingAndFalse
+
+  # Default values.
+  name: ""
+  badge: ""
+  keydown: null # null will be ignored by handlerStack (so it's a safe default).
+  keypress: null
+  keyup: null
+
+  constructor: (options={}) ->
+    Mode.modes.unshift @
+    extend @, options
```

Comment:
This is a strange and very loose contracted. I think it will be much more understandable if we keep options around as its own object, and delegate to it when needed (e.g. keyup invokes option.keyup), or we explicitly assign the properties we expect options to have to `this` (e.g. @keyup = options.keyup).